interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,469 @@
+// apparmor_parser to fail loading the profile. For this situation, we use the
+// PathPattern interface to validate the pattern.
+//
+// Besides that, we are also excluding the `@` character, which is used to mark
+// AppArmor variables (tunables): when generating the profile we lack the
+// knowledge of which variables have been defined, so it's safer to exclude
+// them.
+var (
+	whatRegexp  = regexp.MustCompile(`^/[^"@]*$`)
+	whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^"@]*$`)
+)
+
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9]+$`)
+
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
+	types      []string
+	options    []string
+}
+
+func parseStringList(mountEntry map[string]interface{}, fieldName string) ([]string, error) {
+	var list []string
+	value, ok := mountEntry[fieldName]
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "%s" must be an array of strings (got %q)`, fieldName, value)
+		}
+		for i, iface := range interfaceList {
+			valueString, ok := iface.(string)
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "%s" element %d not a string (%q)`, fieldName, i+1, iface)
+			}
+			list = append(list, valueString)
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return mountAttrTypeError
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return mountAttrTypeError
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if !whatRegexp.MatchString(what) {
+		return fmt.Errorf(`mount-control "what" attribute is invalid: must start with / and not contain special characters`)
```

Comment:
I'll accept the suggestion and remove the "is invalid:" bit to save some characters

Reply:
Ah, it's because I changed it in a different way: I think you accidentally commented on this one, thinking that it was about the "where" attribute. But this error is on the "what" attribute, for which we don't have the `$SNAP_DATA` or `$SNAP_COMMON` restriction.